hazelcast/src/main/java/com/hazelcast/internal/cluster/MemberInfo.java

```diff
@@ -142,6 +159,7 @@ public void readData(ObjectDataInput in) throws IOException {
     public void writeData(ObjectDataOutput out) throws IOException {
         out.writeObject(address);
         UUIDSerializationUtil.writeUUID(out, uuid);
+        UUIDSerializationUtil.writeUUID(out, cpMemberUUID);
```

Comment:
To leverage versioned serialization and make this class rolling-upgrade compatible, it should implement `Versioned` marker interface. Also, append new fields at the end (not in the middle) and read/write them conditionally based on `in/out.getVersion()` being >= `Versions.CURRENT_CLUSTER_VERSION`.
Since `MemberInfo` is not an `Operation` but is included as field in `Operation`s, in order for versioned serialization to work, it must be written with `read/writeObject` methods. Usually this is the case, but there have been some bugs in the past due to outer classes serializing fields by invoking directly `fieldInstance.read/writeData`. So  please confirm that whenever `MemberInfo` is serialized as field of another class, it is serde'd with `read/writeObject` methods.

Reply:
Many thanks for the detailed comments, Vasilis!
I've checked that the `MemberInfo` is not a field of any class. The only fields are a collection of `MemberInfo` that is serialized via `SerializationUtil` or `read/writeObject` by collection iterating.

Have checked the rolling upgrades functionality using Docker